### PR TITLE
SONARJAVA-1871 Expand semantic API to cover parameterized types

### PR DESCRIPTION
**Pull request by m-g-sonar**

### CODE REVIEW

File: java-frontend/src/main/java/org/sonar/java/model/JType.java [hunk 1]
```diff
@@ -25,14 +25,16 @@
...
 import java.util.Objects;
 
-final class JType implements Type, Type.ArrayType {
+final class JType implements Type, Type.ArrayType, Type.ParameterizedType {
 
   final JSema sema;
   final ITypeBinding typeBinding;
 
   private final String fullyQualifiedName;
+  private List<Type> typeArguments;
```
Comment (Godin) [inline java-frontend/src/main/java/org/sonar/java/model/JType.java]: Let's add a comment like in `JSymbol`

```
/**
 * Cache for {@link #typeArguments()}.
 **/
```

File: java-frontend/src/main/java/org/sonar/java/model/JType.java [hunk 2]
```diff
@@ -224,4 +226,24 @@ private static boolean isParameterizedOrGeneric(ITypeBinding typeBinding) {
...
+    return typeBinding.isParameterizedType()
+      // when diamond operator is not fully resolved by ECJ, there is 0 typeArguments, while ECJ
+      // knows it is a Parameterized Type
+      && typeBinding.getTypeArguments().length > 0;
+  }
+
+  @Override
+  public List<Type> typeArguments() {
+    if (typeArguments == null) {
+      ITypeBinding[] typeArgs = typeBinding.getTypeArguments();
```
Comment (Godin) [inline java-frontend/src/main/java/org/sonar/java/model/JType.java]: Let's extract the body of `if` into a method `convertTypeArguments`.

File: java-checks/src/main/java/org/sonar/java/checks/CollectionInappropriateCallsCheck.java [hunk 3]
```diff
@@ -121,14 +121,21 @@ private static String message(Type actualMethodType, Type checkedMethodType, Typ
...
+      return type.name() + ((Type.ParameterizedType) type).typeArguments().stream()
         .map(Type::name)
         .collect(Collectors.joining(", ", "<", ">"));
     }
     return type.name();
   }
 
+  private static List<Type> typeArguments(Type type) {
+    if (type.isParameterized()) {
+      return ((Type.ParameterizedType) type).typeArguments();
```
Comment (Godin) [inline java-checks/src/main/java/org/sonar/java/checks/CollectionInappropriateCallsCheck.java]: Removal of `Type.ParameterizedType` allows removal of this method.

File: java-frontend/src/main/java/org/sonar/plugins/java/api/semantic/Type.java [hunk 4]
```diff
@@ -159,6 +161,25 @@
    */
   Type erasure();
 
+  /**
+   * Check if the current type is a parameterized type or not.
+   * A parameterized type can be casted to a {@link ParameterizedType}
```
Comment (Godin) [inline java-frontend/src/main/java/org/sonar/plugins/java/api/semantic/Type.java]: `ParameterizedType` was removed :wink:

File: java-checks/src/main/java/org/sonar/java/checks/CollectionInappropriateCallsCheck.java [hunk 5]
```diff
@@ -145,8 +145,8 @@ private static Type getMethodOwnerType(MethodInvocationTree mit) {
   }
 
   private static Type getTypeArgumentAt(Type parametrizedType, int parametrizedTypeIndex) {
```
Comment (Godin) [inline java-checks/src/main/java/org/sonar/java/checks/CollectionInappropriateCallsCheck.java]:
```suggestion
  private static Type getTypeArgumentAt(Type type, int index) {
```

**Comment by m-g-sonar** [comment]
Should be good this time @Godin ! I also added tests for UnknownType and a new documentation update for the new API.

**Pull request merged**